near/kycdao-ntnft/src/lib.rs

```diff
@@ -1,21 +1,66 @@
 use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
-use near_sdk::collections::{LookupMap, LazyOption};
+use near_sdk::collections::{LookupMap, LazyOption, UnorderedMap};
 use near_sdk::*;
-use near_sdk::env::keccak256;
+use near_sdk::env::{keccak256, block_timestamp};
 use near_contract_standards::upgrade::Ownable;
 use near_contract_standards::ntnft::{Token, NTNFT, TokenId};
 use near_contract_standards::ntnft::metadata::*;
 
+use serde::{Serialize, Deserialize};
+
 type MintAuthorizationCode = u32;
 
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize)]
+pub struct Status {
+    /// shows if the token is revoked by the issuer
+    pub is_revoked: bool,
+    /// expiry timestamp
+    pub expiry: Option<u64>,
+}
+
+impl Status {
+    pub fn is_valid(&self) -> bool {
+        let expired = match self.expiry {
+            Some(exp) => exp <= block_timestamp(),
+            None => false,
+        };
+        !expired && !self.is_revoked
+    }
+}
+
+impl Default for Status {
+    fn default() -> Self {
+        Status {
+            is_revoked: false,
+            expiry: None,
+        }
+    }
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct OldKycdaoNTNFT {
+    tokens: NTNFT,
+    metadata: LazyOption<NTNFTContractMetadata>,
+    next_token_id: u128,
+    mint_authorizer: AccountId,
+    /// tracks if token minting is authorized, stores metadata temporarily
+    authorized_token_metadata: LookupMap<Vec<u8>, TokenMetadata>,
+}
+
 #[near_bindgen]
 #[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
 pub struct KycdaoNTNFT {
     tokens: NTNFT,
     metadata: LazyOption<NTNFTContractMetadata>,
     next_token_id: u128,
     mint_authorizer: AccountId,
-    authorized_token_metadata: LookupMap<Vec<u8>, TokenMetadata>, // Track if token minting is authorized, store metadata
+    /// tracks if token minting is authorized, stores metadata temporarily
+    authorized_token_metadata: LookupMap<Vec<u8>, TokenMetadata>,
+    /// stores status for authorized (but not yet minted) tokens temporarily
+    authorized_statuses: UnorderedMap<Vec<u8>, Status>,
+    /// stores status for minted tokens
+    token_statuses: UnorderedMap<TokenId, Status>,
```

Comment:
So this is a map stored at contract level for statuses then. Just curious - would it have been possible to store the status at the NTNFT token_statuses level?

Reply:
Yes, but the NTNFT Struct is in our fork of the near-contract-standards, so it meant to be a generic standard and not contain any project specific info. Of course it's not merged and it's not a real standard, but I wanted to maintain this separation, that's why it is on contract level.

Reply:
Ah of course, that makes sense. 👍🏽